### PR TITLE
Remove async blob queue from Request Manager

### PR DESCRIPTION
**Pull request by henridf**
Similar to #5167, but for blobs.

### CODE REVIEW

File: beacon_chain/sync/request_manager.nim [hunk 1]
```diff
@@ -273,41 +271,76 @@ proc requestManagerBlockLoop(rman: RequestManager) {.async.} =
...
+        for worker in workers:
+          if worker.finished() and not(worker.failed()):
+            inc(succeed)
+
+        debug "Request manager blob tick", blobs_count = len(fetches),
+             succeed = succeed,
+             failed = (len(workers) - succeed),
+             sync_speed = speed(start, finish)
+
+      except CancelledError as exc:
```
Comment (tersec) [inline beacon_chain/sync/request_manager.nim:343]:
```suggestion
      except CancelledError:
```

File: beacon_chain/sync/request_manager.nim [hunk 2]
```diff
@@ -273,41 +271,76 @@ proc requestManagerBlockLoop(rman: RequestManager) {.async.} =
...
+
+        await allFutures(workers)
+        let finish = SyncMoment.now(uint64(len(fetches)))
+
+        var succeed = 0
+        for worker in workers:
+          if worker.finished() and not(worker.failed()):
+            inc(succeed)
+
+        debug "Request manager blob tick", blobs_count = len(fetches),
```
Comment (tersec) [inline beacon_chain/sync/request_manager.nim]: The formatting mix here of `blobs_count = len(fetches),` being on the same line as the `debug`, while the rest being individual one per line, separately, probably isn't ideal -- if they all fit in the same line, great, but they should probably all be either consistently on the same line as the `debug` or consistently not on the same line as `debug`.

File: beacon_chain/sync/request_manager.nim [hunk 3]
```diff
@@ -273,41 +271,76 @@ proc requestManagerBlockLoop(rman: RequestManager) {.async.} =
...
 
-      let start = SyncMoment.now(0)
+  var fetches: seq[BlobIdentifier]
+  for blobless in rman.quarantine[].peekBlobless():
 
-      for i in 0 ..< PARALLEL_REQUESTS:
-        workers[i] = rman.fetchBlobsFromNetwork(idList)
+    # give blobs a chance to arrive over gossip
+    if blobless.message.slot == wallSlot and delay < waitDur:
+      debug "Not handling missing blobs as early in slot"
```
Comment (tersec) [inline beacon_chain/sync/request_manager.nim]:
```suggestion
      debug "Not handling missing blobs early in slot"
```

**Review by tersec — APPROVED** [review]

File: beacon_chain/sync/request_manager.nim (hunk 1 above)
Comment (henridf) [reply beacon_chain/sync/request_manager.nim:343]: 8c7617be2ec12f4d9ea87235d1aeb71ddd315d90

File: beacon_chain/sync/request_manager.nim (hunk 2 above)
Comment (henridf) [reply beacon_chain/sync/request_manager.nim]: 8c7617be2ec12f4d9ea87235d1aeb71ddd315d90

File: beacon_chain/sync/request_manager.nim (hunk 3 above)
Comment (henridf) [reply beacon_chain/sync/request_manager.nim]: 8c7617be2ec12f4d9ea87235d1aeb71ddd315d90

**Comment by henridf** [comment]
Test failure is "Geth 0 failed to start" (was green in a previous commit)

**Comment by tersec** [comment]
> Test failure is "Geth 0 failed to start" (was green in a previous commit)

Indeed. https://github.com/status-im/nimbus-eth2/issues/4575#issuecomment-1644488964

Restarted relevant Jenkins CI task/run.

File: beacon_chain/sync/request_manager.nim [hunk 4]
```diff
@@ -69,7 +68,6 @@ proc init*(T: type RequestManager, network: Eth2Node,
   RequestManager(
     network: network,
     # TODO remove this queue and poll the quarantine directly
```
Comment (arnetheduck) [inline beacon_chain/sync/request_manager.nim]:
```suggestion
```

File: beacon_chain/sync/request_manager.nim [hunk 5]
```diff
@@ -268,46 +266,82 @@ proc requestManagerBlockLoop(rman: RequestManager) {.async.} =
...
 
-      debug "Request manager blob tick", blobs_count = len(idList),
-                                         succeed = succeed,
-                                         failed = (len(workers) - succeed),
-                                         queue_size = len(rman.inpBlobQueue)
+proc requestManagerBlobLoop(rman: RequestManager) {.async.} =
+  while true:
+  # TODO This polling could be replaced with an AsyncEvent that is fired
+  #      from the quarantine when there's work to do
+    await sleepAsync(POLL_FREQUENCY)
```
Comment (arnetheduck) [inline beacon_chain/sync/request_manager.nim]: right, let's change to `POLL_INTERVAL` or even period - can never get these right the first time ;)

File: beacon_chain/sync/request_manager.nim [hunk 6]
```diff
@@ -268,46 +266,82 @@ proc requestManagerBlockLoop(rman: RequestManager) {.async.} =
...
+         blk=blobless.root,
+         indices=rman.blobQuarantine[].blobIndices(blobless.root),
+         kzgs=len(blobless.message.body.blob_kzg_commitments)
+      for idx in missing.indices:
+        let id = BlobIdentifier(block_root: blobless.root, index: idx)
+        if id notin fetches:
+          fetches.add(id)
+    else:
+      # this is a programming error should it occur.
+      warn "missing blob handler found blobless block with all blobs"
```
Comment (arnetheduck) [inline beacon_chain/sync/request_manager.nim]: probably want to log some relevant information about the request here - sounds like the kind of thing that would happen after an async race condition

**Comment by henridf** [comment]
`jenkins/nimbus-eth2/linux/x86_64 ` failure is 'Geth 0 failed to start'

**Comment by tersec** [comment]
> `jenkins/nimbus-eth2/linux/x86_64 ` failure is 'Geth 0 failed to start'

https://github.com/status-im/nimbus-eth2/issues/4575#issuecomment-1660878454

**Pull request merged**